### ISSUE TITLE
Calling all contributors - Issues and changes for gmail.js going forward

### ISSUE THREAD

**Issue by josteink** [KartikTalwar/gmail.js#514]
Hey everyone. 

I'm seeing some troubles up ahead. Action *is* required.

## Executive summary

There's issues ahead and I don't foresee any way to resolve them in a fully backwards compatible way. 

I have a proposal for how we go about dealing with these issues, but they mean taking `gmail.js` in a slightly new direction and I'd like some input on those suggestions.

### Context - the new Gmail, the new IDs

Google recently unveiled their "new Gmail" which was more than a GUI-change, it also includes a new data-layer and XHR-format.

What seemed obvious from the first second of testing `gmail.js` on the new gmail was that this was a fundamentally breaking change.

Our strategy *so far* has been to try to patch up the things we see which are reported broken, and try to keep the existing `gmail.js` API-surface unchanged.

In the process of introducing a new data-later, Google has also introduced some new ID-formats: One for threads and one for emails (whereas these used to be the same ID and same format for both in "old" Gmail).

We've tried to the best of our efforts to make `gmail.js` not just a library, but also a forward compatibility-shim which shields the extension writer from having to deal with "old" or "new" Gmail. If that can be done, that's absolutely marvellous and something we should strive for.

But I'm starting to have my doubts.... But why?

### A reality check: the current gmail.js codebase will stop working early 2019

The current gmail.js codebase is 100% tied to legacy IDs, using the old Gmail XHR endpoints.

In some internal/classified documentation provided to various extension-writers, Google has informed us that these API endpoints (and thus any support for legacy IDs) will stop working in "early 2019".

That means that if we keep the codebase as is and do nothing, all `gmail.js`-based code may start failing 4 months from now.

Ladies and gentlemen: The clock is ticking!

### The XHR problem

The biggest and most obvious problem we have right now is that: We're still having issues trying to perform XHR-requests to obtain Gmail-data in new Gmail.

If we cannot use XHR to obtain this data, we will have to resort to fallback methods (either DOM or OAuth).

This particularly impacts `api.get.email_data()` (and all derived/dependant functions).

While there's lots of the info provided by that function can now be provided reasonably reliable through the DOM (which only works if the email is visible), there's absolutely no way we can make this function return the same data, in the same format, with the same granularity and same guarantees once support for legacy IDs is dropped.

So if we can't "fix" that function in a compatible way... We'll have to make a new function whose return-data can be guaranteed, even after legacy IDs (and possible XHR requests) are shut down.

At first I thought about just adding a new `api.dom.email_data()` function and calling it a day...

But then I started thinking about it a bit more deeply, and also considered our other big problem.

### The function/ID ambiguity problem

We have lots of functions with names such as "email_data". We have lots of functions accepting a simple "ID"-parameter.

It's very ambiguous what these functions except and provide. Will they provide email-data? Does this function required a thread-ID or an email-ID?

In the past we didn't have to make that distinction. But now we must.

Now we must be explicit about "I want this info about this thread" and "I want that info about this email". And then I think it makes sense that our functions are explicit about that too: both in naming and return-types.

But making these changes to a published API is not possible in a compatible manner.

### And finally: The reliability problem

Right now, and especially after 2019 enters, as an extension-developer, I can use `gmail.js` in my extension, and at runtime all my API calls will look good. With Typescript I can even verify that I'm using the published API 100% correctly at compile-time!

But while the "build" is good, while the extension initialization is good... There's actually no way for me to know if this code will actually *work* on new Gmail or not without testing.

*With current Gmail.js there is no way to make a guaranteed-to-work extension for new Gmail.*

You have to use the API as provided, hope it works and invest heavily in manual testing. Is that really considered acceptable in 2018?

Basically the API is telling extension-developers "yes, this will be fine", when it may in fact NOT be so. The API is lying.

## Pulling it all together: the proposal

So we know now that:

* We're living on borrowed time. Action required.
* We may lose some functionality (specifically anything XHR-related), and we there *is* a chance we won't be able to make up for it.
* Keeping the current API-surface backwards compatible is not realistic. Trying to do so, will at best lie to extension-authors about the what they can expect returned and when.

In response to that, my proposal is pretty simple:

While we keep the old `Gmail`-class as is (no need to create havoc!), going forward we create a new GmailJs-class (`Gmail2`?), a new API, in which we only expose functionality we know works with new Gmail.

That is, if you as an extension author use `Gmail2` and try to use functionality not supported by new Gmail, your code will either not compile (TypeScript) or cause runtime failures because the methods are not there.

Basically, the new API should not lie.

For the time being this class can simply wrap up call to the existing `Gmail`-class we have, but since this is new code we're writing, we can use this occasion to sort out other issues too:

* The thread/email ambiguity problem. We can now create explicit functions for threads and emails. We can enforce correct ID-format without breaking existing code.
* We can modernize the codebase: Upgrade to ES2015 for all new code. This will give us proper classes, and a structure tooling can understand and support.

Other things which are tempting to at the same time (and really, has to be done at the same time since it affects the API surface):

* Replace callbacks with promises where it makes sense.
* Ditch the jQuery-dependency. It's 2018 and queryMediaSelectorAll works really nice. (And with TypeScript, handling DOM-objects is actually better supported than jQuery).
* Possibly base the new class (classes?) on Typescript, unless there's any major objections.
* Create new, smaller classes instead of a new huge monolith.

### Phew! That was a lot of stuff

Yeah it was. And I'll avoid going into more details for now, but that's roughly it.

What I'm interested in knowing is what you guys think about this proposal. Does it seem reasonable? Or are there things you object to?

cc: All major and recent contributors. @KartikTalwar, @shashikiran-im, @mheinzler , @hankypanky666, @phyxolog, @SuperPaintman, @brentkelly, @madhugb, @Navega, @michi88, @wearhere, @GeneralZero, @avzabr, @neomarbassani, @Capo93

**Comment by davojta** [comment]
I've been working on patching Gmail js internally in our enterprise, and I see additional issues:
*  Gmail lib needs to be injected on the Gmail page itself no be used from content script
* a lot of methods can work fine with dom/selectors and there is no need to inject them to the page
* a lot of magic happens with hacking into internal Gmail API by monkey patching XHR request. That part should be separated and well tested - ideally with autotests. And it is still a question is it reliable to hack internal API.
* there is one more similar lib - https://www.inboxsdk.com/

**Comment by davojta** [comment]
@josteink   I also think about gmailjs refactoring and i am agree with your points and proposals ! 

How is it possible to help with to start that project ?

**Comment by josteink** [comment]
> * Gmail lib needs to be injected on the Gmail page itself no be used from content script
> * a lot of methods can work fine with dom/selectors and there is no need to inject them to the page

I don't really have a big problem injecting the script into the page itself (I need access to the regular DOM anyway), but I can see how some people find this burdonsome.

> a lot of methods can work fine with dom/selectors and there is no need to inject them to the page

Especially if they (from their point view) don't really need it.

But if some portions of Gmail.js can be run in a content page and others not... How does a developer know which functions are safe where?

Unless we can get a good answer for that, I think injecting the script into the page is still the best option.

If we however split the monolith, create smaller dedicated classes... It would probably be possible to use some of these classes directly in the content-page without things getting messy.

> a lot of magic happens with hacking into internal Gmail API by monkey patching XHR request.

We don't really monkey-patch as much as we intercept. [And that part is actually tested](https://github.com/KartikTalwar/gmail.js/blob/master/test/test.tools.js#L8-L49) (not to mention all kind of bad things would start happening immediately if this part failed).

> there is one more similar lib - https://www.inboxsdk.com/

Yes. But it's closed source, proprietary and commercial. And for those reason I (and probably others) do not have much experience with it.

**Comment by josteink** [comment]
>  I also think about gmailjs refactoring and i am agree with your points and proposals !

Good to hear someone agree with my basic design and wants to contribute. That's really appreciated :)

I think a good place to start would be to create an overview of what is currently *known* and tested to fully work on new Gmail. That would include:

* data-operations
* dom-operations
* GUI tools
* etc

If someone could compile a list of those, we can then look over these functions and see which of those are "sensitive" to the whole new vs old ID situation, because obviously we will want to handle that too as early as possible.

Based on this data we should be able to create a initial Gmail2-class which provides wraps these safely. At that point it may or may not make sense to consider the "final" API design. Maybe it makes more sense to throw something together and see how it looks and feels?

IMO we should take that discussion once we get there.

Sounds good?

**Comment by davojta** [comment]
I like the idea to start with something small and see how it will go / will look. I hope to return to the topic during the next days to show something or just create a list with grouped functionality to start with.

**Comment by KartikTalwar** [comment]
Hey all, just catching up to this - I fully support this direction and have no objections. On top of upgrading the codebase to use promises and not jquery, would be great to start separating the functions that support v2 of gmail rewrite

**Comment by davojta** [comment]
I have the following list for first implementation
![image](https://user-images.githubusercontent.com/893650/44323968-ab23d980-a45c-11e8-90bd-ad0f5995819b.png)

**Comment by josteink** [comment]
That looks like a good start, although I'm sure we do have more things supported.

Among other things I know should work reliably are:

* api.dom.thread()
* api.dom.email()
* api.dom.compose() and related functions
* api.get.user_email()
* api.tools.add_modal_window && api.tools.remove_modal_window

**Comment by hankypanky666** [comment]
What do you think about add to new project ts and rx(for chaining action streams)?

**Comment by andreiv05** [comment]
Hello, I'm developing an extension for Gmail and I'm curently trying to integrate gmail.js. First of all, kudos to all the contributors and maintainers, this looks like a great project. Currently, I am using only DOM Actions/Observers, but in the future I might integrate XHR observers too. I think that the documentation of this project is quite old(like, is it really possible to inject jquery.js from the console with CSP enabled?) and, for example,  there is no documentation for `gmail.dom.compose` objects. As I am currently using them, I think that I can write a few examples(something like the  `gmail.dom.email` examples). Looking to help, just give me some start advices.

Thanks,
Andrei

**Comment by hankypanky666** [comment]
> Can you just prototype some pseudocode which demonstrates how a regular event-driven GmailJS would differ from an RX-based solution?

yes, you are rigth RX may confuse contributors, better use Promises

**Comment by josteink** [comment]
Maybe I’ve missed something, but to my knowledge promises and RX handle different kind of usecases and shouldn’t really overlap:

* RX is for a stream of observables, from zero to many occurrences. 
* Promises are for things which are guaranteed (promised!) to occur, but which will always occur exactly once. 

This promises could make sense for data api-calls, while RX would make more sense for events. 

Again: please let me know if I’m missing something here.

**Comment by hankypanky666** [comment]
rx gives us more control over the promises we can start/stop/retry on errors, merge/concat streams, and so on with the help of operators

**Comment by davojta** [comment]
On Wed, Aug 22, 2018 at 4:58 PM Jostein Kjønigsen <notifications@github.com>
wrote:

> Maybe I’ve missed something, but to my knowledge promises and RX handle
> different kind of usecases and shouldn’t really overlap:
>
>    - RX is for a stream of observables, from zero to many occurrences.
>
> It is a idea of the lib. In app you can handle async flow with it.

>
>    - Promises are for things which are guaranteed (promised!) to occur,
>    but which will always occur exactly once.
>
> Exactly! We need something for events, agree. But for me current approach
is okey too.

>
>
> This promises could make sense for data api-calls, while RX would make
> more sense for events.
>
From my experience with RX it is quite big lib (like old fat jQuery) - if
we can to use only few operators and tree shake unused ones - we definitely
should use rx!

P S inboxsdk is use rx internally )

>

**Comment by cboivin80** [comment]
Hello, i would like to contribute also, but i am new on github ... not in Javascript :-). Our business also use gmail.js and we need to pass to the new version.

It's there a good tutorial to start on good base, to know how contribute on github ?

**Comment by cboivin80** [comment]
Finally i found everything i need to contribute!

**Comment by josteink** [comment]
Good good. Let us know if you need anything from the rest of us, and we’ll try to help out.

**Comment by josteink** [comment]
Hey everyone. I got some bad news.

We've received news from Google that the changes they've made which complicates XHR-interaction with new Gmail are deliberate changes made to prevent XHR forgery (like Gmail.js) does.

As far as I can tell they're not going to make any exceptions for us, and that means that if your extension depends on something you can't get from *observing XHR* or *working with the DOM*, you *will* have to migrate your extension to Gmail's official OAuth-based API.

As a timeline, expect XHR-based functions in their current form (like `api.get.email_data()`) to completely stop functioning sometime early in 2019 with no way for us to fix it.

Knowing that, please plan ahead as you see fit. Sorry I can't provide any better options.

**Comment by josteink** [comment]
Hey everyone. I got some bad news.

We've received news from Google that the changes they've made which complicates XHR-interaction with new Gmail are deliberate changes made to prevent XHR forgery (like Gmail.js) does.

As far as I can tell they're not going to make any exceptions for us, and that means that if your extension depends on something you can't get from *observing XHR* or *working with the DOM*, you *will* have to migrate your extension to Gmail's official OAuth-based API.

As a timeline, expect XHR-based functions in their current form (like `api.get.email_data()`) to completely stop functioning sometime early in 2019 with no way for us to fix it.  Knowing that, please plan ahead as you see fit. Sorry I can't provide any better options.

That said, if this is not an issue for you... Please help us build the future proof GmailJS2 version which we will have to make work within these constraints :smiley: 

I'll be back in a more active developer-like position in about a week or two and hope I can start making some real progress then.

**Comment by pttdev** [comment]
"observing XHR or working with the DOM"? Why they prevent us from working with the DOM? Is that possible? I understand the XHR part but we can still inject javascript to interact with the DOM. Am I right?

**Comment by josteink** [comment]
No you're reading that wrong. Let me try to rephrase it:

The only things we will be able to do in the future is:

* observing XHR
* working with the DOM

Anything outside that (like actually creating our own XHR requests) will be unsupported and there probably won't be a way for us to fix that shortcoming. And *then* you will have to use Google's OAuth API instead.

Is that more clear?

**Comment by pttdev** [comment]
Thank you for clarify that. One more question, will this effect the send_message hook? in case we need to modify the body before sending?

**Comment by josteink** [comment]
No idea. I haven’t done anything like that myself. You’ll have to test on your own.

**Comment by josteink** [comment]
Hey everyone.

I'm *finally* back in mode to work on GmailJS and I've been doing some research, and at last I'm ready to make some progress on this issue!

For real this time. No more "next week".
🎆 🍾 🎉 🎈 

So welcome to another "oh its a long josteink post" post 😄 

## First a disclaimer

I'm making this comment mostly as a *user* of GmailJS in the context of my company's GMailJS-based WebExtension.

That means my interests are primarily aligned with keeping my company's extension working, and not so much with lifting "all" of GmailJS to the level you might want.

Therefore I will ensure the parts I need gets lifted. I know this may seem like selfish abuse of my status as a maintainer, but I only have so much time to work on this, and I really don't have capacity to help everyone else on the parts only they need. Sorry.

Basically: If you want your parts included in the upcoming "GmailJS 2.0", you may have to lift them yourself. Ofcourse I will try to be helpful in aiding you mainline your changes in a way I think is sustainable, but you will probably have to put in some effort yourself.

The only other realistic option for me and my company would be to simply fork and keep our code private. And I would *really* like to avoid that.

I hope you can understand, and that this is OK with everyone. 

## APIs I will work on lifting/porting

Back to the topic at hand. I've tried to map whatever my company uses Gmail.js for to whatever can be supported without forged XHR requests.


As a *library-user*, the things my company depends on in our extension is not a whole lot: 

* api.get.user_email()
* api.observe.on() - "compose", "load", "view_email", "load_email_menu"
* api.observe.after("send_message")
* api.get.localization()
* api.get.thread_id()
* api.get.email_data() - but only some of the data returned
* api.helper.get.email_source_pre()
* api.dom.email()
* api.dom.email.attachments()
* api.dom.compose()
* api.tools.add_attachment_button(...)
* api.tools.add_compose_button(...)
* api.tools.add_modal_window(...)

You can safely assume that these functions will be *partially* lifted.

I've tried inspecting the implementation for all of those to map out what they strictly depend on, and to what degree they will be affected by the upcoming changes.

As far as I can tell, by rewriting relevant parts of GMailJS to use a XHR interception and DOM inspection *only*, my company should be mostly safe, with a few exceptions.

Through DOM alone, I can almost make out everything we need `api.get.email_data()` for, with only these exceptions: 

* subject
* date

I can clearly observe those data through interception XHR requests, so it should be possible to store them and cache these data for later use though.

## Legacy IDs live on... for now

As an interesting observation, even new Gmail uses *legacy IDs* for obtaining a the raw mime data (through "Show original").

So even in New GMail (and thus GMailJS 2), we will have to keep Legacy IDs around for the sake of supporting mime-data fetching.

I'm thinking simply adding a GmailMailID type class which embeds *all* the known ID-formats into one pack, which can then (with type-safety!) be thrown around to various methods, without consumers having to having worry about ID-formats at all.

Which brings us to the next point...

## GMailJS backwards compatibility - email IDs

Making email/thread ID a complex type should simplify quite a lot in *green-field* code, but will obviously require rewriting certain portions in existing code-bases.

The benefit (for typescript-users at least) should be that this change will be a *type-error* which can be caught at compile time instead of passing various format IDs, all in strings, only to discover bugs at runtime.

## GMailJS backwards compatibility - email_data()

`email_data()` will also not be able to return the exact same data it has before, and especially for emails not already fetched by Gmail. We won't be able to trigger a fetch, so that kind of action may end up entirely unsupported.

Unfortunately it goes without saying `email_data()` can't remain 100% backwards compatible. Depending on my findings, I may not even aim for that.

I can't even begin to imagine the amount of extensions and lines of code which will be needed changed due to this... But there's really not much I can do about it.

## Reconsidering Gmail backwards compatibility

I've also observed that *Google is slowly removing the ability to go back to "old" GMail*.

For my "regular" gmail-account this feature is already gone. For my GSuite account, I'm about to be forced onto New GMail. I'm guessing the "old" gmail will soon be removed from this account too.

I don't know if old GMail is fully gone before or after I've done the job of lifting GmailJS to work reliably on New GMail. But there is a reasonable chance it may be.

And if so, complicating the design and investing in backwards-compatibility against something which won't exist clearly seems like a waste of time.

For me this means I may consider compatibility with "old" GMail non-essential in GMailJS 2. My first design and revision will certainly omit this, and only later will I add it, if it still look needed.

I can see how this may be slightly controversial, but at least I will be open about my ambitions and aims here. If you find this highly objectionable, let me at least know... And we will just have to work things out one way or the other.

## Typescript

I like TypeScript. I *really* like TypeScript. And lots of other people do, and I don't think this is a "controversial" or "niche" WebExtension development-language any more.

I've mentioned a possible GMailJS 2 implementation being based on TypeScript without seeing any objections so far. In the absence of such objections, I'll move forward with code based on TypeScript. Thanks for your approval 😄 

## A final note

That's a lot of information. It may be a lot to digest. There's quite a few changes there.

I'd really like to avoid upsetting the small community we have, but at the same time I see the need to take the action required to help keep what we have built working. I think it's a tough balance, and I may make missteps.

Please take time to think about these changes, think about the big picture (like I've at least tried to do), and let me know what you think.

**Comment by josteink** [comment]
I've started doodling, and after some consideration I've decided that the simplest and least intrusive way to do this is to actually start a new project from scratch, using current gmail.js as a dependency.

You can find it here: https://github.com/josteink/gmail.ts

By doing things this way:

* I'm not breaking gmail.js.
* I preserve all the proven code we have and do a gradual migration as new code gets added/updated.
* I can make fundamental changes in the API as needed, in the process exploit this to work better with modern toolkits and platforms. 

Please note that currently this should only be considered an experiment, and not something I'm committing to 100%.

A gmailjs compatablity layer *might* be added, but I'm not making any promises. At any rate, this won't be a drop-in replacement.

If it works out, or I'm happy after I'm done doodling, chances are it will stay pretty much as is. So if you have some input about the design and how you think things could be done differently, letting me know at this stage is probably a good idea.


Again: Feedback would very much be welcome.

**Comment by josteink** [comment]
I've received no feedback and that's OK.

Quick update:

* I got stuck in technical issues w.r.t. building it as a library (not as an application) and having the build process export correct/usable JS & TypeScript annotations.
* I wanted to implement way more intense XHR-monitoring (to replace forged XHR-requests), and this forced me to do most of the work in Gmail.js's observer-code (not in Gmail.ts)
* In the end gmail.ts was a lot more work than I had expected, with a lot less impressive payoff than I had anticipated.

So in the end I've implemented support for the things I need in some new, not-backwards compatible APIs added to `gmail.js`. And if I've already done the work there, why bother with `gmail.ts` at all then?

For those interested, this new code lives in a a new feature-branch: https://github.com/KartikTalwar/gmail.js/tree/feature/prefetch-data

**I plan on merging this to master within a week or two**, so if you want to provide feedback that's your window. After this I will consider it a published, public API, and making changes to it will be a lot harder.

Quick summary of the changes:

* When user is using new Gmail, snoop into *all* XHR communications by default. 
* Parse and cache all XHR requests which contains email-thread related data. 
* Provide a new `api.new.get.*` namespace, for new gmail only.
* The most important function in this namespace is `api.new.get.email_data()` which does not rely on forged XHR requests, but instead use the in-memory cached data from XHR-interception. Performance is therefore (obviously) much better.
* Using the cache imposes some limitations: It wont be able to get email data about *any* email... 
* However Gmail *will* issue XHR-requests to fetch (and prefetch) all emails in the inbox or whatever folder or view you are in, so if the user at any point has seen an email or its listing, `api.new.get.email_data()` should be able to provide the data you need, and instantly.

**Comment by davojta** [comment]
@josteink i have go through your code and it looks good to me )

it is hard to build such a thing as gmailjs from scratch without significant efforts, thank you a lot for try! 

On our project, we reserve some time (not so much, like 1 week) to help you with migration and I think we will spend that time to audit the current code base for breaking changes since 2019 ( i mean migration from legacy id to new one)

**Comment by josteink** [comment]
And merged.

The main readme/docs needs to be updated, but for those who wants to give this a try without fetching git feature-branches, this is now published as version 0.8.0.

Closing issue as *main* problem seems to be solved.

To everyone who contributed and provided feedback: Thanks. It's appreciated!

**Comment by josteink** [comment]
A final FYI, I've now messed around with the new code in a production-extension and found it somewhat unergonomic, and needlessly hard to port in places.

As a result I've made some breaking changes w.r.t. the shape of data returned from `api.new.get.email_data()` to make migration easier.

These changes can be found in the latest official package (0.9.0), and at this point I'm going to consider the API stable unless someone objects .

If you're unsure if code you wrote for 0.8.0 still works, I highly advise using TypeScript as this should catch all breaking errors.
